### ISSUE TITLE
Allow (Proguard) obfuscation of the Quarkus app in fast-jar format 

### ISSUE THREAD

**Issue by devprojectsme** [quarkusio/quarkus#25602]
### Description

Dear Quarkus Team, 
there is [this issue](https://github.com/quarkusio/quarkus/issues/6034), where people have requested some support for obfuscation. Dear [Guillaume Smet](https://github.com/gsmet) has suggested, that it could be done as an external project, and has closed that issue "for now". However, as I was not able to find any of that initiative I would at least like to summarize here the two problems I have confronted with so far, as a person trying to learn and migrate existing microservices to Quarkus.
There are mainly 2 problems, if you decide to obfuscate the Quarkus app released in the "fast-jar" format (quarkus default, as of 1.5 see [here](https://github.com/quarkusio/quarkus/wiki/Migration-Guide-1.12#fast-jar-as-default)): 

a) The first issue is, that Quarkus takes the initiative and tries to speed-up the boot time by generating the implementations  of [io.quarkus.arc.InbjectableBean](https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/InjectableBean.java) beans specification. It is all great, but it uses internally the Class.forName() method, to load those classes then.Moreover stores that part under _quarkus-app/quarkus/generated-bytecode.jar_ Hence the obfuscator has to be "smart enough" to be able to sind the string occurrences in the generated classes, in two different JARs(under _quarkus-app/app_ and under _quarkus-app/quarkus_). To overcome this, you have to let those CDI beans unobfusctaed,or move the actual processing logic in a separate beans and obfuscate those. 

b) The new format utilizes the [QUARKUS_APPLICATION_DAT](https://github.com/quarkusio/quarkus/blob/5c120b1e5c8e4c4bd587581137edd2762472b52f/independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/QuarkusEntryPoint.java#L21) file for internal serialization of the important app information in own binary format. 
This is an issue, as the obfuscator changes the class/package names, but those then can not be found, as they stay unchanged in the QUARKUS_APPLICATION_DAT.

Those two issues require the developer to move the code that has to be obfuscated to some other classes, which might be tedious considering the vast features of REST, [SmallRey  Fault Resilience](https://quarkus.io/guides/smallrye-fault-tolerance), and further  available as Quarkus Extensions. 

### Implementation ideas

Not sure, as this would lead to changes of current architecture. One most trivial implementation would be to allow changing the  QUARKUS_APPLICATION_DAT externally, with some provided mapping.txt. For that one could use the [MappingPrinter](https://github.com/Guardsquare/proguard/blob/ccab80e755d2af4547a8af0821c2c50e32122470/base/src/main/java/proguard/obfuscate/MappingPrinter.java) and [MappingReader](https://github.com/Guardsquare/proguard/blob/6dfa0ca00e090d8343f7f8359d5f2fb6850fb69b/base/src/main/java/proguard/obfuscate/MappingReader.java) Classes from Proguard repo. However that is only an idea so far, and the internal arc, structure of the resulting JAR, and actually the legal legislations of using those classes as the dependency would be needed. As an update in mappingsPrinter or reader would need an update in the new independent project for "changing" the QUARKUS_APPLICATION_DAT file externally.

**Comment by geoand** [comment]
Thanks for reporting.

To be honest, this is not something we want to do. Obfuscation can most likely be done without any of the issues you mention for `fast-jar`, if the application is built with `quarkus.package.type=legacy-jar` as that type does not have separate jar nor does it keep extra metadata in the form of a binary file.

**Comment by devprojectsme** [comment]
Dear @geoand ,
thanks for your answer and suggestion. 

I have tried to use the `quarkus.package.type=legacy-jar`, but even there Quarkus still generates a lot of stuff for startup, especially the Configuration classes in package `io.quarkus.runtime.generated`. I have tried to trace and to debug it, and have then bumped into the problem, that the new framework for SmallRye config generation uses the ':' colon (see [here](https://github.com/quarkusio/quarkus/blob/2.8.0.Final/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java#L1124)) for building the method names. That specific character is not a valid symbol for JavaIdentifierPart(see [here](https://docs.oracle.com/javase/specs/jls/se7/html/jls-3.html#jls-3.8)) which then prevents the IDE from debugging that source. 

All of this makes the migration process for someone learning quarkus harder, so if possible it would be great to have one example for the obfuscation in [Quarkus-quickstarts](https://github.com/quarkusio/quarkus-quickstarts) project.

**Comment by geoand** [comment]
Hi,

>  but even there Quarkus still generates a lot of stuff for startup

By a lot of stuff, what do you mean? Quarkus does **not** generate **any** classes at runtime, all classes are generated at build time and contained in the `*-runner` jar (when using the `legacy-jar` package).

**Comment by devprojectsme** [comment]
Hi @geoand , thanks for responding.
 
By "a lot of stuff" I mean the content of the `io.quarkus.runtime.generated`, specifically the `io.quarkus.runtime.generated.Config` class with the `readConfig(ConfigSourceProvider var1)` method, that then in turn calls all of the generated methods like: `static void initGroup$io$quarkus$runtime$logging$LogConfig(SmallRyeConfig var0, StringBuilder var1, Object var2)` that read dynamically the configuration. 

I do understand, that those classes are generated in the build-time( the generator does use the colon in the method names as I have described above). 

My current issue is, that even though my Proguard config contains the instructions to keep all of the classes for Quarkus,Vertx, Netty, etc. the obfuscated jar can not start due to the exception by initializing the config file, that occurs in those generated classes. 

As I can not debug it directly(see the issue with the colon above), I have to find the reason why that is happening. Currently I am using the [BTrace](https://github.com/btraceio/btrace) script, that allows me a glimpse into the executions of the  `readConfig(ConfigSourceProvider var1)` method.

Thus at the moment I have an issue, that the non-obfuscated app version stars and runs without any problem, the obfuscated one fails to start due to reading the config through the generated classes.  Here the example of the exception in the generated code with the stacktrace (from an obfuscated app), so it fails before starting the app classes itself:

<details><summary>Stacktrace</summary><blockquote>
`
...
ERROR: Failed to start application (with profile prod)
java.util.NoSuchElementException
        at io.quarkus.runtime.configuration.NameIterator.prevPos(NameIterator.java:157)
        at io.quarkus.runtime.configuration.NameIterator.getPreviousSegment(NameIterator.java:290)
        at io.quarkus.runtime.generated.Config.rtGetEnclosing:log:LogConfig#filters:*:CleanupFilterConfig#ifStartsWith(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtGetEnclosing:log:LogConfig#filters:*:CleanupFilterConfig#ifStartsWith(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtGetEnclosing:log:LogConfig#filters:*:CleanupFilterConfig#ifStartsWith(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtGetEnclosing:log:LogConfig#filters:*:CleanupFilterConfig#ifStartsWith(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtGetEnclosing:log:LogConfig#filters:*:CleanupFilterConfig#ifStartsWith(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtParseKey:quarkus:log:filter:*:if-starts-with(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtParseKey:quarkus:log:filter:*(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtParseKey:quarkus:log:filter(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtParseKey:quarkus:log(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtParseKey:quarkus(Unknown Source)
        at io.quarkus.runtime.generated.Config.rtParseKey(Unknown Source)
        at io.quarkus.runtime.generated.Config.readConfig(Unknown Source)
        at io.quarkus.deployment.steps.RuntimeConfigSetup.deploy(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
        at io.quarkus.runtime.Application.start(Application.java:101)
        at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:103)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:67)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:41)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:120)
        at io.quarkus.runner.GeneratedMain.main(Unknown Source)
`
</blockquote></details>

**Comment by devprojectsme** [comment]
Hi, 
I have not done the whole custom build yet, as there are so many different projects to set up. I have tried to change and "release" locally only the `io.quarkus.quarkus-core-deployement` and then use it, within my POM.xml. Then I have kept changing my tracing script, until have found one discrepancy between the obfuscated and un-obfuscated jar. Namely, some of the generated methods keep their signature, but have some changed source, i.e. see the following screenshot:
<details><summary>decompiled source</summary><blockquote>
    <img src="https://i.ibb.co/JpmKGRj/Strange-Decompiled.png" alt ="https://i.ibb.co/JpmKGRj/Strange-Decompiled.png"/>
</details>

My Proguard config contains the  instructions to neither optimize nor to shrink the code, only to obfuscate our code, keeping the code from the libs as mentioned above.  

Hence currently I am no longer sure it is the Quarkus issue once using the `quarkus.package.type=legacy-jar` , and would then keep digging and let you know once I find out why that is happening.
